### PR TITLE
feat(sdk/elixir): rework on Elixir codegen

### LINKED ISSUE #5901: 🐞 Elixir codegen didn't get cleanup when generate failure
### What is the issue?

https://discord.com/channels/707636530424053791/1120503349599543376/1162417244878409768

### Log output

_No response_

### Steps to reproduce

_No response_

### SDK version

Elixir 0.8.8

### OS version

macOS

### LINKED ISSUE #6310: 🐞 Elixir: strange experimentalWithAllGPUs function name
### What is the issue?

See https://hexdocs.pm/dagger/Dagger.Container.html#experimental_with_all_gp_us/1. The function name should be `experimental_with_all_gpus`, `experimental_with_all_gp_us`

### Dagger version

dagger v0.9.4

### Steps to reproduce

_No response_

### Log output

_No response_

### PR DESCRIPTION
**Pull request by wingyplus**
This is a big change but let me explain to catch up easier.

## Codegen moved out from the SDK

Previously, the dagger codegen was embedded in the SDK, this caused a problem in that it could not be run when the codegen generated a bad code. This change introduces a new Elixir module called `dagger_codegen` which's responsible for generating code and uses the `escript` to build a binary instead of the mix task. 

The way to fetch introspection is now split out of the codegen. This way can reduce the compilation annoying issue, described in https://github.com/dagger/dagger/issues/5901, but the issue still happens when using `hack/make sdk:elixir:generate`. 

## `sdk:elixir` tasks improvement

The `hack/make sdk:elixir:*` task is now upgraded to use Elixir 1.16 with Erlang/OTP 26. The `sdk:elixir:generate` will not produce a bad code when its failure.

## Nicer code-generation

Previously, the codegen used Elixir quote to produce code but it's not suitable for our use case, the codegen is a bit hard to read and needs to deal with complex AST structure. The new codegen uses a new approach by using EEx as a template and moving some complex code-generation to the renderer module.

The typespec and documentation also improved.

Here is comparison

0.9.7

Screenshot 2567-02-01 at 23 33 01
Screenshot 2567-02-01 at 23 34 22

From this PR

Screenshot 2567-02-01 at 23 33 58
Screenshot 2567-02-01 at 23 34 59

## Fix issues in the codegen

- This temporarily fixes issue #6310 about the acronym word format from camel to snake case.
- Fix #5901 by do not export a bad code during run `sdk:elixir:generate` as mentioned above.
- Fix removed APIs not get cleanup when run `sdk:elixir:generate`.

## Cleanup

- Eliminate `Nestru` out of the codegen since we change how to query list object like `envVariables`. 
- Takedown the codegen in `lib/dagger/codegen`.

## Improvement on the next PR

- Improve test coverage on codegen.
- Make codegen easier to write. Currently, some of it is living inside the template but some of it looks complex living inside the module.
- Make required arguments more safety with guards.

### CODE REVIEW

**Comment by wingyplus** [comment]
Found the `env_variables` API broken (#6594). We need to rework the simple object also. 😢

File: sdk/elixir/lib/dagger/gen/container.ex
Comment (wingyplus) [inline sdk/elixir/lib/dagger/gen/container.ex:144]: Any function that return a list of object are breaking changes since version 0.9.7 but didn't notice user about this breaking changes. Do we need to add a changelog about this?

File: internal/mage/sdk/elixir.go
Comment (wingyplus) [inline internal/mage/sdk/elixir.go:144]: It would be nice if mage tasks (or incoming ci module) could inject an introspection JSON file and let the codegen use that file.

**Comment by wingyplus** [comment]
The PR size is quite big now. So I decided to stop on this and continue on a new PR for incoming improvements.

**Comment by jedevc** [comment]
> The PR size is quite big now. So I decided to stop on this and continue on a new PR for incoming improvements.

@wingyplus what do you want to do about this PR? Should it be reviewed/merged, or do you want to split it into smaller ones? Either works for us.

**Comment by wingyplus** [comment]
> > The PR size is quite big now. So I decided to stop on this and continue on a new PR for incoming improvements.
> 
> @wingyplus what do you want to do about this PR? Should it be reviewed/merged, or do you want to split it into smaller ones? Either works for us.

Review and merge is the way I prefer. For the mage task changes, I'll update on #6843 to make it up-to-date if this PR merge before that PR. 🙇

**Comment by wingyplus** [comment]
Rebased.

**Comment by gerhard** [comment]
This is on my list for today.

If we can resolve the conflicts before then, merging this will be straigthforward.

Want to take a look too @gmile?

File: sdk/elixir/dagger_codegen/lib/dagger/codegen/introspection/types/type_ref.ex
Comment (gerhard) [inline sdk/elixir/dagger_codegen/lib/dagger/codegen/introspection/types/type_ref.ex:31]: 👍

File: sdk/elixir/lib/dagger/core/query_builder.ex
Comment (gerhard) [inline sdk/elixir/lib/dagger/core/query_builder.ex:21]: Did you mean to remove this now, or in a follow-up PR?

**Review by gerhard — APPROVED** [review]
This is a HUGE PR @wingyplus 😱 My apologies for not getting to it earlier.

This looks great to me 💪 I am keen to get it moving & so that you can follow-up with the rest.

In case @gmile and @helderco want to check it out too, please do, otherwise I'm excited for the follow-ups 🙌

---

FWIW, we - especially @jedevc - are communicating upcoming releases via https://github.com/dagger/dagger/milestones, so you can see what release is coming up, and when.

**Review by gmile — COMMENTED** [review]
Amazing work! 👏🏅 Looking forward to having this merged!

File: sdk/elixir/lib/dagger/core/query_builder.ex
Comment (wingyplus) [reply sdk/elixir/lib/dagger/core/query_builder.ex:21]: I need to tackle it on the PR while I rework on the query builder.

File: internal/mage/sdk/elixir.go
Comment (helderco) [inline internal/mage/sdk/elixir.go:235]: Simplification:
```suggestion
		WithDirectory(mountPath, util.Repository(c).Directory(elixirSDKPath)).
```

File: sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/formatter.ex
Comment (helderco) [inline sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/formatter.ex:125]: clarify?

File: sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/formatter.ex
Comment (helderco) [inline sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/formatter.ex:24]: Must make sure to handle possible reserved words in the language, I couldn't spot this being done.

Maybe an "end" function argument, or a "do" function name. Anywhere where there could be a conflict, most importantly that would lead to a syntax error, but can also do it for soft keywords as well (shadowing built-ins). Example in Python:

https://github.com/dagger/dagger/blob/795a73595d8dd0c14bc9b71e6a137852e7e0e9ac/sdk/python/src/dagger/_codegen/generator.py#L345-L346

With modules, codegen is open to many end user choices, it's not just core. But even in the core API, I've already seen us choosing a different name than the one desidered due to a conflict with the codegen in an experimental SDK. It was easier to choose a different name then, but it's better to make this a requirement for SDKs going forward.

File: sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/renderer.ex
Comment (helderco) [inline sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/renderer.ex:74]: We have a convention that deprecation reason messages can reference a field (most likely the field that should be used instead), wrapped in backticks (e.g.,  ``use `newField` instead``). The reason for this is that SDKs can extract whatever's inside the backticks from the deprecation reason and re-format it according to its own naming conventions. 

Example in Python:

https://github.com/dagger/dagger/blob/795a73595d8dd0c14bc9b71e6a137852e7e0e9ac/sdk/python/src/dagger/_codegen/generator.py#L664-L673

https://github.com/dagger/dagger/blob/795a73595d8dd0c14bc9b71e6a137852e7e0e9ac/docs/docs-graphql/schema.graphqls#L2142

https://github.com/dagger/dagger/blob/795a73595d8dd0c14bc9b71e6a137852e7e0e9ac/sdk/python/src/dagger/client/gen.py#L5696

**Review by helderco — APPROVED** [review]
Ok to iterate on follow-ups, so approving now. 🚀

File: sdk/elixir/dagger_codegen/mix.exs
Comment (wingyplus) [reply sdk/elixir/dagger_codegen/mix.exs:8]: Changing it to 1.14 would align with the main package. 👍

File: sdk/elixir/scripts/fetch_introspection.exs
Comment (wingyplus) [reply sdk/elixir/scripts/fetch_introspection.exs:4]: Yes. :)

File: sdk/elixir/dagger_codegen/mix.exs
Comment (wingyplus) [reply sdk/elixir/dagger_codegen/mix.exs]: Fixed on the current commit.

File: sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/formatter.ex
Comment (wingyplus) [reply sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/formatter.ex:125]: My bad. 😂

File: sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/renderer.ex
Comment (wingyplus) [reply sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/renderer.ex:74]: Currently, I've done field conversion by transform content in backquote into snakecase. But the hard part is to link to referenced function in Elixir, we need the arity which's the number of argument of that function. So it's now become a little bit tricky that we need to traverse introspection to find the exact number. 😢 

 I think it would be great to follow up this issue on the next PR.

Comment (wingyplus) [reply sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/renderer.ex:74]: Reference to issue https://github.com/dagger/dagger/issues/6939

File: sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/formatter.ex
Comment (wingyplus) [reply sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/formatter.ex:24]: I see. Adding `_` suffix to align with Python is interesting. Add the issue https://github.com/dagger/dagger/issues/6940 to follow up later. 🙇

File: internal/mage/sdk/elixir.go
Comment (wingyplus) [reply internal/mage/sdk/elixir.go:144]: > That's done in modules, more specifically via the SDK runtime module. The server provides the introspection JSON and so, codegen needs to be able to read from that.

Yeah, I see it from the module runtime. So I thought it would be nice to have this included in generate task on non-support module SDK (rust, java, elixir, etc.). 

> An external way to get it is to:
> 
> ```
> dagger -m github.com/helderco/daggerverse/codegen@v0.1.2 call introspect as-json -o introspection.json
> ```

Ahh, so we can calling the function using `cliBinPath` to download introspection and then using it to generate code. Nice! Do we have a plan to include it in new CI module (#6843)?

File: sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/formatter.ex
Comment (wingyplus) [reply sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/formatter.ex]: Fixed.

File: internal/mage/sdk/elixir.go
Comment (wingyplus) [reply internal/mage/sdk/elixir.go]: Fixed.

File: sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/renderer.ex
Comment (helderco) [reply sdk/elixir/dagger_codegen/lib/dagger/codegen/elixir_generator/renderer.ex:74]: Oh, now I see that's replaced in all descriptions, not just deprecations. We've avoided referencing other fields in the description in the core API because of the wrong naming convention in other languages, but I think that's no longer the case. It's easy to creap in. 

We've limited this to deprecation reasons because it's more common in field descriptions to want something in backquotes other than being a field name. Doesn't seem like there's a case like this today though, but that can change. It's smart to only match valid identifiers, but I can't rule out a name in backquotes that is a valid identifier, but meant for something else other than an API field.

Changing the casing is enough. When I said "reference", I wasn't referring to anything special in code, even though Python does generate a prefix that allows it to be hyperlinked correctly in the reference docs. But the convention here is just to rename, having IDE autocomplete in mind.

**Comment by wingyplus** [comment]
I think this PR is in a good shape now. So I'm gonna merge this now.

Thank you @gerhard @helderco @gmile for helping review this PR.

File: internal/mage/sdk/elixir.go
Comment (helderco) [reply internal/mage/sdk/elixir.go:144]: No plan to put it in the CI module, but you can do it for Elixir 🙂

**Pull request merged**
